File: app/models/music/bound_note.rb
# frozen_string_literal: true

module Music
  # Represents a note on the piano at a specific frequency (and in a specific octave). Can be thought of as a specific
  # fret on a guitar or a specific key on a piano. Useful for representing actual music that might be printed out
  # as sheet music or a guitar tab or what have you.
  class BoundNote
    def self.symbolic(string_or_note)
      if string_or_note.respond_to?(:note?) && string_or_note.note?
        if string_or_note.bound?
          string_or_note
        else
          raise "Can't create BoundNote from UnboundNote #{string_or_note.inspect}"
        end
      else
        symbol = NoteSymbol.parse(string_or_note.to_s)
        frequency = FrequencyConverter.frequency_for_symbol(symbol)
        new(frequency, symbol)
      end
    end

    attr_reader :symbol, :frequency

    def initialize(frequency, symbol = nil)
      @frequency = frequency
      @symbol = if symbol.nil?
        FrequencyConverter.symbol_for_frequency(frequency)
      elsif symbol.is_a?(NoteSymbol)
        symbol
      else
        NoteSymbol.parse(symbol)
      end
    end

    include Comparable
    def <=>(other)
      if other.class == self.class
        frequency <=> other.frequency
      else
        -1
      end
    end

    def ==(other)
      if other.class == self.class
        frequency == other.frequency
      end
    end
    alias_method :eql?, :==

    def apply_interval(interval)
      new_octave, new_letter_index = (octave * 12 + semitones_above_c + interval.semitones).divmod(12)
      new_symbol = if symbol.letter.include? 'b'
        UnboundNote::SEMITONES_TO_FLAT_NOTES.fetch(new_letter_index)
      else
        UnboundNote::SEMITONES_TO_SHARP_NOTES.fetch(new_letter_index)
      end

      symbol = NoteSymbol.new(new_symbol, new_octave)
      frequency = FrequencyConverter.frequency_for_symbol(symbol)
      self.class.new(frequency, symbol)
    end

    def hash
      @frequency.hash
    end

    def symbol_without_octave
      @symbol.letter
    end

    def bound?
      true
    end

    def note?
      true
    end

    def unbind
      UnboundNote.new(symbol_without_octave)
    end

    def semitones_above_c
      UnboundNote::NOTES_TO_SEMITONES.fetch(@symbol.letter)
    end

    def semitones_above_c4
      (octave - 4) * 12 + semitones_above_c
    end

    def octave
      @symbol.octave
    end
  end
end

File: app/javascript/music/bound_note.ts
// Represents a note on the piano at a specific frequency (and in a specific octave). Can be thought of as a specific
// fret on a guitar or a specific key on a piano. Useful for representing actual music that might be printed out
// as sheet music or a guitar tab or what have you.
import { FrequencyConverter } from "./frequency_converter";
import { INote } from "./i_note";
import { Interval } from "./interval";
import { NoteSymbol } from "./note_symbol";
import { UnboundNote, NotesToSemitones, SemitonesToFlatNotes, SemitonesToSharpNotes } from "./unbound_note";

export class BoundNote implements INote {
  public static fromString(noteString: string) {
    const symbol = NoteSymbol.parse(noteString);
    const frequency = FrequencyConverter.frequencyForSymbol(symbol);
    return new BoundNote(frequency, symbol);
  }

  public static sorter = (a: BoundNote, b: BoundNote) => a.frequency - b.frequency;

  public symbol: NoteSymbol;
  public octave: number;
  public symbolWithoutOctave: string;
  public semitonesAboveC: number;
  public semitonesAboveC4: number;

  constructor(public frequency: number, symbol?: NoteSymbol) {
    if (!symbol) {
      this.symbol = FrequencyConverter.symbolForFrequency(frequency);
    } else {
      this.symbol = symbol;
    }
    this.symbolWithoutOctave = this.symbol.letter;
    this.octave = this.symbol.octave;
    this.semitonesAboveC = NotesToSemitones[this.symbol.letter];
    this.semitonesAboveC4 = (this.octave - 4) * 12 + this.semitonesAboveC;
  }

  public bound() {
    return true;
  }

  public unbind() {
    return new UnboundNote(this.symbolWithoutOctave);
  }

  public applyInterval(interval: Interval): BoundNote {
    const totalSemitones = this.octave * 12 + this.semitonesAboveC + interval.semitones;
    const newOctave = Math.floor(totalSemitones / 12);
    const newLetterIndex = totalSemitones % 12;

    const newSymbol = this.symbol.letter.includes("b") ?
                        SemitonesToFlatNotes[newLetterIndex] :
                        SemitonesToSharpNotes[newLetterIndex];
    const symbol = new NoteSymbol(newSymbol, newOctave);
    const frequency = FrequencyConverter.frequencyForSymbol(symbol);
    return new BoundNote(frequency, symbol);
  }

  public equivalent(note: BoundNote) {
    return note.frequency === this.frequency;
  }
}
